web/packages/teleterm/src/services/tshdEvents/index.ts

```diff
@@ -118,100 +122,104 @@ async function createServer(
-// need the UI to react to the events. We cannot create the service in the UI code because this
-// would mean that the service would need to cross the contextBridge. This is simply impossible
-// because the service is fed custom gRPC classes which can't be passed through the contextBridge.
-//
-// Instead, we create an event emitter and expose subscribeToEvent through the contextBridge.
-// subscribeToEvent lets UI code register a callback for a specific event. That callback receives
-// a simple JS object which can freely pass the contextBridge.
-//
-// # Async behavior
-//
-// The callback can return a promise. The service will not return a response until all callbacks
-// resolve. This lets us model behavior where tshd calls the Electron app and then blocks until it
-// receives a response, in case the Electron app needs to do some work before we want to unblock
-// tshd.
-//
-// If any of the callbacks return an error, the service will return that error immediately, without
-// waiting for other listeners.
+/**
+ * createService creates a service for the tshd events server. It sets up the preload part of the
+ * service and expects the UI to call setupTshdEventContextBridgeService to add the browser part of
+ * the service which interacts with the UI.
+ *
+ * See the JSDoc for TshdEventContextBridgeService for more details.
+ */
 function createService(logger: Logger): {
   service: apiService.ITshdEventsServiceServer;
-  subscribeToTshdEvent: SubscribeToTshdEvent;
+  setupTshdEventContextBridgeService: (
+    listener: TshdEventContextBridgeService
+  ) => void;
 } {
-  const emitter = new Emittery();
+  let contextBridgeService: TshdEventContextBridgeService;
 
-  const subscribeToTshdEvent: SubscribeToTshdEvent = (eventName, listener) => {
-    emitter.on(eventName, listener);
+  const setupTshdEventContextBridgeService = (
+    newListener: TshdEventContextBridgeService
+  ) => {
+    contextBridgeService = newListener;
   };
 
-  const service: apiService.ITshdEventsServiceServer = {
-    relogin: (call, callback) => {
-      const request = call.request.toObject();
-
-      logger.info('Emitting relogin', request);
+  /**
+   * processEvent is a helper function encapsulating common operations done before and after sending
+   * the event to the browser through the context bridge.
+   *
+   * The last argument is a function which maps the response object (received from the browser
+   * through the context bridge) to a class instance (as expected by grpc-js).
+   */
+  function processEvent<
+    RpcName extends keyof apiService.ITshdEventsServiceServer,
+    Request extends protobuf.Message,
+    Response extends protobuf.Message
+  >(
+    rpcName: RpcName,
+    call: grpc.ServerUnaryCall<Request, Response>,
+    callback: (error: Error, response: Response) => void,
+    mapResponseObjectToResponseInstance: (
+      responseObject: ReturnType<Response['toObject']>
+    ) => Response
```

Comment:
This doesn't technically type check anything, even when replacing `Request` and `Response` with more specific types from `ITshdEventsServiceServer` and I don't really know how to make this work.

You can have a gRPC handler like this and TypeScript won't say a word:

```typescript
    sendNotification: (call, callback) =>
      processEvent(
        'sendNotification',
        call,
        callback,
        () => new api.ReloginResponse() // ❌ Wrong response class.
      ),
```

Reply:
Oh my bad, this actually does work. It just seems like it doesn't for RPCs with empty responses, because the object versions of those responses are just empty objects.